xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,264 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+/*
+ListNET is adapted from the paper:
+Cao, Zhe, et al. "Learning to rank: from pairwise approach to listwise approach."
+Proceedings of the 24th international conference on Machine learning. ACM, 2007.
+*/
+
+#include "xapian-letor/ranker.h"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <vector>
+
+using namespace std;
+using namespace Xapian;
+
+ListNETRanker::ListNETRanker() {
+    // default values
+    learning_rate = 0.0001;
+    iterations = 15;
+    cout << "Initializing ListNETRanker with default parameters" << endl;
+}
+
+ListNETRanker::ListNETRanker(double learn_rate, int num_iterations) {
+    learning_rate = learn_rate;
+    iterations = num_iterations;
+    cout << "Initializing ListNETRanker with Learning Rate: " << learn_rate;
+    cout << ", No. of Iterations: " << iterations << endl;
+}
+
+ListNETRanker::~ListNETRanker() {
+}
+
+static double
+calculateInnerProduct(vector<double> parameters, vector<double> fvals) {
+
+	double inner_product = 0.0;
+
+	for (int i = 0; i < int(fvals.size()); ++i){
+
+		inner_product += parameters[i] * fvals[i];
+	}
+
+	return inner_product;
+}
+
+// From Theorem (6) in paper.
+static vector< vector<double> >
+initializeProbability(vector<FeatureVector> feature_vectors, vector<double> & new_parameters) {
+
+    int list_length = feature_vectors.size();
+
+    // probability distribution, y is ground truth, while z is predict score
+    vector<double> prob_y;
+    vector<double> prob_z;
+    double expsum_y = 0.0;
+    double expsum_z = 0.0;
+
+    for (int i = 0; i < list_length; i++) {
+        expsum_y += exp(feature_vectors[i].get_label());
+        expsum_z += exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()));
+    }
+
+    for (int i = 0; i < list_length; i++) {
+        prob_y.push_back(exp(feature_vectors[i].get_label())/expsum_y);
+        prob_z.push_back(exp(calculateInnerProduct(new_parameters,feature_vectors[i].get_fvals()))/expsum_z);
+
+    }
+
+    vector< vector<double> > prob;
+    prob.push_back(prob_y);
+    prob.push_back(prob_z);
+
+	return prob;
+}
+
+// Equation (3) in paper.
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, vector< vector<double> > prob) {
+
+	vector<double> gradient(feature_vectors[0].get_fcount(),0);
+	int list_length = feature_vectors.size();
+
+	vector<double> prob_y(prob[0]);
+	vector<double> prob_z(prob[1]);
```

Comment:
Are these intermediate variables needed? I think it's just copying a `vector<double>` for no reason.